Fix SyncBatchNorm for empty inputs

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#74944 Fix SyncBatchNorm for empty inputs**

fixes #36530

Prior to this commit, SyncBatchNorm crashes with the following
error message.

```
File "..../torch/nn/modules/_functions.py", line 17, in forward
    mean, invstd = torch.batch_norm_stats(input, eps)
RuntimeError: cannot reshape tensor of 0 elements into shape [0, 3, -1] because the unspecified dimension size -1 can be any value and is ambiguous
```

This PR adds a dedicated branch to handle empty inputs. When a process
recieves empty inputs, it will set its local `mean`, `invstd`, and `count`
to zero, and participate in the `all_gather` collective communications in
the forward pass. Then `mean` and `invstd` with zero count will be
filtered out before computing global mean and invstd. In the backward
pass, it also participate in the `all_reduce` communication with zero
tensors to unblock its peers.

Differential Revision: [D35273409](https://our.internmc.facebook.com/intern/diff/D35273409)

mrshenli committed:
[WIP] Fix SyncBatchNorm for empty inputs

TODO:
1. avoid copying count_all to CPU if possible
2. it's not crashed any more, but the output is nan

Next step will try to move the fix to the CUDA kernel of
`batch_norm_gather_stats_with_counts` accordingly

[ghstack-poisoned]

datumbox commented on test/distributed/test_c10d_common.py:
```diff
@@ -806,6 +806,28 @@ def fut_then(fut):
+        )
+
+        model.train()
+        x = torch.zeros((1 if self.rank != 0 else 0, 2, 11, 13), dtype=torch.float32, device=self.rank)
```
We are passing an input with a single value here (all zeroes) which means that the standard deviation will be 0 so it's inverse would be `inf` which might be the source of your `nan`. 

Perhaps we should change this to
```suggestion
        x = torch.randn((1 if self.rank != 0 else 0, 2, 11, 13), dtype=torch.float32, device=self.rank)
```

datumbox commented on torch/nn/modules/_functions.py:
```diff
@@ -16,17 +16,21 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+            combined = torch.cat([mean, invstd, count], dim=0)
+        else:
+            # for empty input, directly set all stats to 0
+            combined = torch.zeros(2 * num_channels + 1, device=input.device)
```
Nit:
```suggestion
            combined = torch.zeros(2 * num_channels + 1, dtype=input.dtype, device=input.device)
```

datumbox commented on torch/nn/modules/_functions.py:
```diff
@@ -69,8 +78,10 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+        if input.numel() > 0:
+            return torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
+        else:
+            return torch.empty_like(input)
```
I've tested something similar before but placed this at the beginning of `forward`. Unfortunately placing it at the end has the same issue, the training hangs. I've tried this combined patch (with all the comments included) on real data & cluster. After a few iterations, when we eventually sent empty data to one of the devices the training freezes. I've sshed to all the nodes of the cluster and I see that all the GPUs that received data report 100% utilization while one of them reports 0%. 

Am I right to suggest that this `if` statement messes with the graph and causes issues on the backwards step? Perhaps what we want is for `batch_norm_elemt` to actually handle empty input?

mrshenli replied:
>  all the GPUs that received data report 100% utilization while one of them reports 0%.

This means all expect one is waiting for collective communication, but someone one of them didn't join the collective comm. 

> Am I right to suggest that this if statement messes with the graph and causes issues on the backwards step?

Yep, basically the above error prevents that process from joining the collectives. Let me try a fix tonight.

datumbox reviewed (commented):
Thanks for the investigation @mrshenli. I've had a look as well and added a few comments. Let me know your thoughts.

mrshenli commented on torch/nn/modules/_functions.py:
```diff
@@ -53,6 +57,11 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
             mean_all, invstd_all, count_all = torch.split(combined, num_channels, dim=1)
 
+        # remove stats from empty inputs
+        indices = [i for i, x in enumerate(count_all.cpu().tolist()) if x[0] >= 1]
```
This is brilliant! Thanks!

mrshenli committed:
Update on "[WIP] Fix SyncBatchNorm for empty inputs"




[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35273409).

datumbox reviewed (commented):
Thanks for the change @mrshenli. 

Overall the approach looks good to me. I've added minor comments for nits. I'm currently testing this patch on a cluster using real data and it seems that the problem is resolved. If something breaks, I'll let you know.

albanD commented on torch/nn/modules/_functions.py:
```diff
@@ -1,3 +1,4 @@
+from hamcrest import none
```
Well if we land with broken CI, the PR will be reverted. So we definitely want to fix them :D

mrshenli replied:
Thanks for spotting this! Honestly, I don't know why this import was added. Must be me mis-typed sth. Will remove. :)

albanD commented on torch/nn/modules/_functions.py:
```diff
@@ -16,17 +17,25 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+            # C, C, 1 -> (2C + 1)
+            combined = torch.cat([mean, invstd, count], dim=0)
+        else:
+            # for empty input, directly set all stats to 0
```
I'm not sure this comment is accurate. The code works because all inputs for which the count is 0 are filtered out later. Not because the stats contain 0s.

mrshenli replied:
added more detailed descriptions.

albanD commented on torch/nn/modules/_functions.py:
```diff
@@ -16,17 +17,25 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
+            combined = torch.cat([mean, invstd, count], dim=0)
+        else:
+            # for empty input, directly set all stats to 0
+            combined = torch.zeros(
```
Wouldn't something like: `torch.zeros(dtype=input.dtype, device=input.device).expand(2 * num_channels + 1)` also work and reduce the bandwidth that is wasted?
Not sure how the rpc is handling non-contiguous Tensors.

mrshenli replied:
```
torch.zeros(dtype=input.dtype, device=input.device).expand(2 * num_channels + 1)
```

Curious, what bandwidth does the above code save? And why RPC is relevant here?

mrshenli replied:
Oh I see. Not sure if this gonna work. Collectives use ProcessGroup and will call NCCL APIs under the hood. IIRC, NCCL expects contiguous tensors and will directly read `numel()` elements from the memory pointer. Let me double check on that

mrshenli replied:
```
  File "/raid/shenli/pytorch/torch/distributed/distributed_c10d.py", line 2130, in _all_gather_base                                                             
    work = group._allgather_base(output_tensor, input_tensor)                                                                                                   
RuntimeError: Tensors must be contiguous                                                                                                                        
Exception raised from check_gpu_single_tensor at ../torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1227 (most recent call first):  
```

Hit the above error, caused by the following line.  

https://github.com/pytorch/pytorch/blob/835cc66e5dd26db558931b4fe47b45e08a3a09f7/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1226-L1228

albanD replied:
Ok then.
As a side note, I think you should look into that as it is potentially a major bandwidth gain (and if I understand correctly, this is an expensive commodity).

albanD commented on torch/nn/modules/_functions.py:
```diff
@@ -80,45 +97,73 @@ def backward(self, grad_output):
+            num_channels = saved_input.shape[1]
+            if self.needs_input_grad[0]:
+                # launch all_reduce to unblock other peer processes
+                combined = torch.zeros(
```
Same question about expanded Tensor to reduce bandwidth use

albanD commented on torch/nn/modules/_functions.py:
```diff
@@ -80,45 +97,73 @@ def backward(self, grad_output):
+            else:
+                grad_input = None
+
+            grad_weight = torch.zeros(
```
This is not needed. If the result will be just zeros, you can just return None which is interpreted by the autograd as a Tensor full of zeros.

mrshenli replied:
Does this also apply to `grad_input`. If I leave that as None, is it guaranteed that autograd engine will also interpret that as a tensor of zeros?

mrshenli committed:
Update on "Fix SyncBatchNorm for empty inputs"


fixes #36530

Prior to this commit, SyncBatchNorm crashes with the following
error message.

```
File "..../torch/nn/modules/_functions.py", line 17, in forward
    mean, invstd = torch.batch_norm_stats(input, eps)
RuntimeError: cannot reshape tensor of 0 elements into shape [0, 3, -1] because the unspecified dimension size -1 can be any value and is ambiguous
```

This PR adds a dedicated branch to handle empty inputs. When a process
recieves empty inputs, it will set its local `mean`, `invstd`, and `count`
to zero, and participate in the `all_gather` collective communications in
the forward pass. Then `mean` and `invstd` with zero count will be
filtered out before computing global mean and invstd. In the backward
pass, it also participate in the `all_reduce` communication with zero
tensors to unblock its peers.

Differential Revision: [D35273409](https://our.internmc.facebook.com/intern/diff/D35273409)

[ghstack-poisoned]

mrshenli commented on torch/nn/modules/_functions.py:
```diff
@@ -80,45 +97,73 @@ def backward(self, grad_output):
+
+            if weight is None or not self.needs_input_grad[2]:
+                grad_bias = None
+        else:
```
Yep, that's exactly what this PR does.

mrshenli commented on test/distributed/test_c10d_common.py:
```diff
@@ -806,6 +806,101 @@ def fut_then(fut):
+
+        # input does not requires grad
+        x.requires_grad = False
+        self._test_not_nan(model, x)
```
@datumbox @albanD is there a way to test the grad value as well?

If I feed batch size 0 and 2 to the two processes, will it generate the same gradients if they receive batch size of 1 and 1 respective of the same data? I assume no, because `invstd` is no longer guaranteed to be the same in this case?

datumbox replied:
I think I agree. It's not going to be the same gradient because the minibatch statistics will be different in the two cases.

ppwwyyxx replied:
@datumbox  I don't get why you say "the minibatch statistics will be different" - in SyncBatchNorm the minibatch is __all samples__ on all workers, so in @mrshenli 's example it's the same minibatch (2 samples) in both cases.

mrshenli committed:
Update on "Fix SyncBatchNorm for empty inputs"


fixes #36530

Prior to this commit, SyncBatchNorm crashes with the following
error message.

```
File "..../torch/nn/modules/_functions.py", line 17, in forward
    mean, invstd = torch.batch_norm_stats(input, eps)
RuntimeError: cannot reshape tensor of 0 elements into shape [0, 3, -1] because the unspecified dimension size -1 can be any value and is ambiguous
```

This PR adds a dedicated branch to handle empty inputs. When a process
recieves empty inputs, it will set its local `mean`, `invstd`, and `count`
to zero, and participate in the `all_gather` collective communications in
the forward pass. Then `mean` and `invstd` with zero count will be
filtered out before computing global mean and invstd. In the backward
pass, it also participate in the `all_reduce` communication with zero
tensors to unblock its peers.

Differential Revision: [D35273409](https://our.internmc.facebook.com/intern/diff/D35273409)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35273409).

mrshenli committed:
Update on "Fix SyncBatchNorm for empty inputs"


fixes #36530

Prior to this commit, SyncBatchNorm crashes with the following
error message.

```
File "..../torch/nn/modules/_functions.py", line 17, in forward
    mean, invstd = torch.batch_norm_stats(input, eps)
RuntimeError: cannot reshape tensor of 0 elements into shape [0, 3, -1] because the unspecified dimension size -1 can be any value and is ambiguous
```

This PR adds a dedicated branch to handle empty inputs. When a process
recieves empty inputs, it will set its local `mean`, `invstd`, and `count`
to zero, and participate in the `all_gather` collective communications in
the forward pass. Then `mean` and `invstd` with zero count will be
filtered out before computing global mean and invstd. In the backward
pass, it also participate in the `all_reduce` communication with zero
tensors to unblock its peers.

Differential Revision: [D35273409](https://our.internmc.facebook.com/intern/diff/D35273409)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35273409).

datumbox reviewed (approved):
LGTM from my side. My tests on real-data show that the issue is fixed.

albanD reviewed (approved):
SGTM
